### PR TITLE
Activate element inside given flow scope using ancestor selection

### LINKED ISSUE #9646: Modify a process instance by activating elements inside a given flow scope
## Description

We want to allow the activation of elements inside a given flow scope instance (i.e. ancestor selection). The flow scope instance is selected by its element instance key. The selected instance could be the flow scope of the element or one of its upper scopes. 

By default, the new element instance is created within an existing flow scope instance (#9640 / #9642). Or, new flow scope instances are created if they aren't active yet (#9643).

But we need the additional option to select the ancestor scope instance:
* if multiple instances of the flow scope are active 
* if the element instance should be created in a new instance of a flow scope

If the selected ancestor scope is the direct flow scope then the new element instance is created within this scope instance. 

For example:
image

If the selected ancestor scope is not the direct flow scope but an upper one (or the process instance itself) then new flow scope instances are created. Create a new flow scope instance for each flow scope between the selected ancestor and the given element. Create a new element instance of the given element within the created flow scope instance.   

For example:
image

**What needs to be done?**
* for each activate instruction:
  * if the selected ancestor flow scope instance is the direct flow scope then create the element instance as in #9642 
  * otherwise, create new flow scope instances under the selected ancestor as in #9643  

**These issues should be completed once ancestor selection is available**
* #9984
* #9990
* #11031
* #11032

**Out of scope**
* active an element inside a multi-instance

**Blocked by**
- #9643

### PR DESCRIPTION
**Pull request by korthout**
## Description

This enables ancestor selection for activate instructions of process instance modification.

When multiple instances of a flow scope of the activate instruction's target element exist, then the engine must somehow choose which of these instances to use as the flow scope of the newly activated element instance. The engine cannot make this choice. It should be up to the user. By providing an ancestor scope key along with the activate instruction, the user can decide this for the engine.

- If the ancestor scope key refers to an element instance that is **the direct flow scope** of the activate instruction's target element, then the engine will create the new element instance directly inside of the selected ancestor.

- If the ancestor scope key refers to an element instance that is **an indirect flow scope** of the activate instruction's target element, then the engine will create a new element instance of each indirect and direct flow scope in between the selected ancestor and the target element.

- If no ancestor scope key is provided, the modification command is rejected like before, with a rejection message clarifying what happened. This does not yet mention the ability to use ancestor selection, as this is meant to be resolved in #11032.

## Related issues

closes #9646 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by korthout** [comment]
@saig0 This PR is not yet ready to be merged, but I'd like to ask you to have a look at [c14d971](https://github.com/camunda/zeebe/pull/11086/commits/c14d9713bf6bc64929dcc84eb88c4010b6c1f0fb). I've attempted to cover a particular part of the code, but I cannot find a way to reach it. I'd be curious to learn your opinion about this part of the code and whether you think it's possible to cover it in a test.

**Comment by saig0** [comment]
@korthout I assume that you can't test the recursive call of `isAncestorOfElementInstance()`. 

I also don't see how to reach the recursive call. It's always the direct flow scope. 

However, I found a case that is currently not covered. In the linked test case, I try to activate the element `C` in one of the four instances of the subprocess `sub`. I set the ancestor key to one of the subprocess instances. The process instance doesn't apply the modification because it failed to determine the instance of the event subprocess `event-sub`. There are two instances of the event subprocess but the ancestor key is none of them.

Please have a quick chat if you have any questions.

**Comment by korthout** [comment]
I think I managed to cover the case. I'd still prefer to discuss this f2f. I'll send an invite

**Comment by korthout** [comment]
@saig0 I've refactored the code for improved understandability, including more specific naming and additional documentation. Please have a look. I recommend looking at the changes per commit because some changes are structural while others are simply renamings

File: engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java [hunk 1]
```diff
@@ -810,6 +817,310 @@ public void shouldActivateElementInInterruptedFlowScope() {
...
+
+    // then
+    final var activatedTaskB =
+        RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+            .withProcessInstanceKey(processInstanceKey)
+            .withElementId("B")
+            .findAny();
+    assertThat(activatedTaskB).describedAs("Expect that task B is activated").isPresent();
+    assertThat(activatedTaskB.get().getValue())
+        .describedAs("Expect that task B is exists inside of flow scope " + ancestorScopeKey)
```
Comment (saig0) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/ModifyProcessInstanceTest.java:879]: :wrench: Typo.

```suggestion
        .describedAs("Expect that task B exists inside of flow scope " + ancestorScopeKey)
```

**Review by saig0 — APPROVED** [review]
@korthout great job. :1st_place_medal: 

I like the result. :+1: And, it was a joy to follow the commits to it. :rocket:

**Comment by korthout** [comment]
bors merge

**Comment by korthout** [comment]
bors retry

**Pull request merged**